ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,363 @@
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the `/workspace/source` mount point for all containers in the build-pod.
+Shipwright Build Controller also adds it's own helper to ["git clone" the
+repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+In the `BuildRun` resource, the Build Controller will inspect for the following annotations:
+
+- `build.shipwright.io/upload.wait`: boolean attribute, only `true` or `false` accepted;
+- `build.shipwright.io/upload.wait.timeout`: sets the timeout;
+
+The idea is to employ annotations as we would employ flags in a command-line application, they inform
+the Build Controller about the desired changes for that particular `BuildRun` instance.
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  annotations:
+    build.shipwright.io/upload.wait: "true"
+    build.shipwright.io/upload.wait.timeout: "30s"
+spec: {}
+```
+
+Therefore when `build.shipwright.io/wait.upload` is present the waiting will take place, and by
+expressing `build.shipwright.io/upload.wait.timeout` the actual timeout is configured.
+
+Only the `BuildRun` object will accept the "flags" to trigger the waiting for user upload, meaning
+this type of approach is not meant to leak towards the parent `Build` resource.
+
+##### Alternative BuildRun API
```

Comment:
not sure about the word "Alternative" here

I was envisioning that we *at least* had the `type: LocalCopy` from the outset, but then initially control the other implementation detail items via annotation.

Reply:
I also think we need to mention either here or somewhere that if we are not doing it at least initially, we need to start thinking about whether opting into `LocalCopy` moves from the `BuildRun` to the `Build` .... "promoting" it if you will.

given the uncertainty we have on several aspects of the implementation details (file ownership, upload method, etc), I like starting off at the `BuildRun` ... feels safer.  Even if we add the field to the `Build` later, the `BuildRun` field could still serve as an override.

But conversely, I think it is likely that when we start getting feedback from users, they are going to at least ask about the selection for local source in the sources array (where the URL in `BuildSource` could be `file://  ....`

So again, even if we are not going there initially with this EP in its current form, so verbiage about keeping an eye toward ^^ seems warranted IMO.

Reply:
Seperate point: yes, with the need to "opt in" via `type: LocalCopy` and/or annotations, at least IMO, we have preserved the rebase image scenario from @SaschaSchwarze0 as I understand it).

Reply:
another clarification occurred to me 

- ideally, we end up in a place like most k8s api where there are just api fields and we don't need annotations to control
- if there is enough uncertainty, of suspicion that the API might have to pivot, an interim combination of new api fields and annotations exist
- so we are just iterating on what the initial balance of new api and annotations is 
- with that, at least for me, `type: LocalCopy` with a  combination of initial annotations is the correct initial balance

Reply:
> not sure about the word "Alternative" here
> 
> I was envisioning that we _at least_ had the `type: LocalCopy` from the outset, but then initially control the other implementation detail items via annotation.

I'm rewording it to explain the comparison between those two approaches instead. Please consider.